examples/rcn.py

```diff
@@ -0,0 +1,390 @@
+# ---
+# jupyter:
+#   jupytext:
+#     formats: ipynb,py:percent
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.11.4
+#   kernelspec:
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# %%
+# %matplotlib inline
+import os
+import time
+
+import jax
+import matplotlib.pyplot as plt
+import numpy as np
+from jax import numpy as jnp
+from jax import tree_util
+from scipy.ndimage import maximum_filter
+from scipy.signal import fftconvolve
+from sklearn.datasets import fetch_openml
+
+from pgmax.fg import graph, groups
+
+os.environ["XLA_PYTHON_CLIENT_ALLOCATOR"] = "platform"
+os.environ["XLA_PYTHON_CLIENT_PREALLOCATE"] = "false"
+
+# %% [markdown]
+# # 1. Load the data
+# %%
+hps, vps = 12, 12
+
+# Use train_size = 100 if you have a gpu with atleast 8Gbs memory.
+# Recommend that jax is installed with cuda enabled for this option.
+train_size = 100
+test_size = 20
+
+
+def fetch_mnist_dataset(test_size: int, seed: int = 5):
+    """Returns test images sampled randomly from the set of MNIST images.
+    Args:
+        test_size: Desired number of test images.
+    Returns:
+        test_set: A list of length test_size containing images from the MNIST test dataset.
+        test_labels: Corresponding labels for the test images.
+    """
+
+    mnist_train_size = 60000
+    num_per_class = test_size // 10
+
+    print("Fetching the MNIST dataset")
+    dataset = fetch_openml("mnist_784", as_frame=False, cache=True)
```

Comment:
From playing with the notebook, the caching does not seem to work. Please investigate and we should make caching work.

Reply:
This seems to be a known issue.
https://github.com/scikit-learn/scikit-learn/issues/18783

Fixed using the suggestions in this thread.